core/templates/dev/head/pages/exploration-player-page/services/fatigue-detection-service.spec.ts

```diff
@@ -0,0 +1,90 @@
+// Copyright 2017 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the 'License');
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an 'AS-IS' BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Unit tests for the Fatigue Detection service.
+ */
+
+require('pages/exploration-player-page/services/fatigue-detection.service.ts');
+
+describe('Failure detection service', function() {
+  beforeEach(angular.mock.module('oppia'));
+
+  var FatigueDetectionService = null;
+  beforeEach(angular.mock.inject(function($injector) {
+    FatigueDetectionService = $injector.get('FatigueDetectionService');
+  }));
+
+  beforeEach(function() {
+    jasmine.clock().uninstall();
+    jasmine.clock().install();
+    jasmine.clock().mockDate();
+  });
+
+  afterEach(function() {
+    jasmine.clock().uninstall();
+  });
+
+  describe('isSubmittingTooFast', function() {
+    it('should return true for 4 submissions in under 10 seconds', function() {
+      FatigueDetectionService.recordSubmissionTimestamp();
+      jasmine.clock().tick(100);
+      FatigueDetectionService.recordSubmissionTimestamp();
+      jasmine.clock().tick(100);
+      FatigueDetectionService.recordSubmissionTimestamp();
+      jasmine.clock().tick(100);
+      FatigueDetectionService.recordSubmissionTimestamp();
+
+      expect(FatigueDetectionService.isSubmittingTooFast()).toBe(true);
+    });
+  });
+
+  describe('isSubmittingTooFast', function() {
+    it('should return false for 4 submissions in over 10 seconds', function() {
+      FatigueDetectionService.recordSubmissionTimestamp();
+      jasmine.clock().tick(100);
+      FatigueDetectionService.recordSubmissionTimestamp();
+      jasmine.clock().tick(100);
+      FatigueDetectionService.recordSubmissionTimestamp();
+      jasmine.clock().tick(10000);
+      FatigueDetectionService.recordSubmissionTimestamp();
+
+      expect(FatigueDetectionService.isSubmittingTooFast()).toBe(false);
+    });
+  });
+
+  describe('isSubmittingTooFast', function() {
+    it('should return false for less than 4 submissions', function() {
+      FatigueDetectionService.recordSubmissionTimestamp();
+      FatigueDetectionService.recordSubmissionTimestamp();
+      FatigueDetectionService.recordSubmissionTimestamp();
+
+      expect(FatigueDetectionService.isSubmittingTooFast()).toBe(false);
```

Comment:
You could also check that this holds at the end of each submission, so that you are testing the "1 submission" and "2 submission" cases as well.

Optional: Perhaps worth having the Jasmine clock tick slightly between the submissions, too? In order to ensure that the `false` isn't happening just because all the submissions are being submitted near-simultaneously and being counted as just a single-time submission somehow. (Trying to make this robust to changes in the possible implementation of FatigueDetectionService.)

Reply:
Seems reasonable to me. Done.